Integrando database NOSQL ao projeto

fpbitencourt opened the pull request:
1. Setup SnnapyDB
2. Persistindo AppState
3. DataManager implementa StateListener e persiste o último estado modificado da app
4. Referência do DataManager centralizado no Application
5. DataManagerTest - whenAppStateChangesOnceTime and whenAppStateChangesSeveralTime
6. Novo Action 'APP_STATE_LOADED' para reduzir o estado inicial da app
8. Jackson para Serealizar e Desserealizar AppState e Gif
9. Implementado equals para as classes AppState e Gif
10. Setado multiDexEnabled, solução intermediária para o limite de métodos no projeto (https://github.com/Catbag/redux-android-sample/issues/56)

fpbitencourt committed:
Integrando database NOSQL ao projeto

1. Setup SnnapyDB
2. Persistindo AppState
3. DataManager implementa StateListener e persiste o último estado modificado da app
4. Referência do DataManager centralizado no Application
5. DataManagerTest - whenAppStateChangesOnceTime and whenAppStateChangesSeveralTime
6. Novo Action 'APP_STATE_LOADED' para reduzir o estado inicial da app
8. Jackson para Serealizar e Desserealizar AppState e Gif
9. Implementado equals para as classes AppState e Gif
10. Setado multiDexEnabled, solução intermediária para o limite de métodos no projeto (https://github.com/Catbag/redux-android-sample/issues/56)

fpbitencourt committed:
debug drone

niltonvasques commented on app/build.gradle:
```diff
@@ -69,6 +84,15 @@ dependencies {
+    compile 'co.trikita:anvil-sdk15:0.4.0'
     compile 'com.android.support:recyclerview-v7:24.0.+'
     compile 'pl.droidsonroids.gif:android-gif-drawable:1.2.2'
+    compile 'com.github.trikita:anvil-recyclerview:1.2'
```
duplicated

niltonvasques commented on app/build.gradle:
```diff
@@ -69,6 +84,15 @@ dependencies {
     compile 'com.github.trikita:anvil-recyclerview:1.2'
     compile 'com.squareup.retrofit2:retrofit:2.1.0'
     compile 'com.squareup.retrofit2:converter-gson:2.1.0'
+    compile 'co.trikita:anvil-sdk15:0.4.0'
```
duplicated

fpbitencourt replied:
Falha no merge. Será que as ferramentas de inspect não poderiam sinalizar isso?

niltonvasques commented on app/src/androidTest/java/br/com/catbag/gifreduxsample/asyncs/data/DataManagerTest.java:
```diff
@@ -0,0 +1,120 @@
+ */
+
+@RunWith(AndroidJUnit4.class)
+public class DataManagerTest {
```
I guess that others methods from DataManager also should be tested.

fpbitencourt replied:
Quais?

niltonvasques replied:
Probably all public methods.

fpbitencourt replied:
Concordo. Boa parte surgiu após puxar da integração com a master, por isso não olhei para eles. Abrirei uma issue para isso.

fpbitencourt replied:
Na verdade, só vejo necessidade de fazer teste do _fetchGifs_

niltonvasques replied:
E qual a garantia temos de que os demais estão ok?

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/asyncs/data/DataManager.java:
```diff
@@ -70,8 +131,36 @@ public void onFailure(Call<RiffsyResponse> call, Throwable t) {
     }
 
-}
+    private class SaveAppStateTask extends AsyncTask<Void, Void, Boolean> {
```
There is a subtle problem with AsyncTasks to perform background operations.

When a fresh AsyncTask is launched, it's auto bind the current activity on screen, and retains it's object even when the activity is finished. Which can leads to memory leaks.

There are several articles talking about the issues with AsyncTasks: 
- http://blog.danlew.net/2014/06/21/the-hidden-pitfalls-of-asynctask/
- http://simonvt.net/2014/04/17/asynctask-is-bad-and-you-should-feel-bad/
- http://bon-app-etit.blogspot.com.br/2013/04/the-dark-side-of-asynctask.html

fpbitencourt replied:
Manolo, li os artigos mas não consegui compreender se ela é sempre _bindada_ à _actitivy_ atual. No caso atual, ela não é explicitamente associada à nenhuma activity. Fiz o teste encerrando a _activity_ atual e depois abrindo novamente enquanto a _asynctask_ executava em background e funcionou normalmente.

niltonvasques replied:
Mas não vai crashar sempre. No entanto, ela fica segurando a activity que estava presente no momento que a async foi criada, mesmo após a activity já ter saido da tela. O que por sua vez gera memory leaks, já que o Garbage Colector não vai poder limpar o objeto da activity. Pra testar essa comportamento especifico da Async deve dá um trabalhozinho.

Veja: 

> Memory leaks
> Because an AsyncTask has methods that run on the worker thread (doInBackground()) as well as methods that run on the UI (e.g. onPostExecute()), it has took keep a reference to it's Activity as long as it's running. But if the Activity has already been destroyed, it will still keep this reference in memory. This is completely useless because the task has been cancelled anyway. ref: http://bon-app-etit.blogspot.com.br/2013/04/the-dark-side-of-asynctask.html

niltonvasques replied:
Isso é potencializado ainda mais, nos casos onde o Android decide recriar a Activity por causa de algum evento ( como por exemplo a rotação de tela). Nesse caso se a Async tiver sido criada antes da rotação, ela vai guardar a referência da Activity anterior, levando novamente a um memory leak.

> Losing your results
> Another problem is that we lose our results of the AsyncTask if our Activity has been recreated. For example when an orientation change occurs. The Activity will be destroyed and recreated, but our AsyncTask will now have an invalid reference to its Activity, so onPostExecute() will have no effect. There is a solution for this. You can hold onto a reference to AsyncTask that lasts between configuration changes (for example using a global holder in the Application object). Activity.onRetainNonConfigurationInstance() and Fragment.setRetainedInstance(true) may also help you in this case.

fpbitencourt replied:
Entendi! É que pra mim não tava claro que a AsyncTask sempre estava vinculada a uma activity. Estranho a documentação do Android não alertar os desenvolvedores para isso. Valeu pela information! Alterarei o código para Thread + Runnable.

niltonvasques replied:
É foda isso manolo. Eu descobri isso recente da pior maneira, debugando uns bugs loucos com fragments e async tasks. :disappointed:

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/asyncs/data/storage/Database.java:
```diff
@@ -1,19 +1,53 @@
  * Created by felipe on 26/10/16.
  */
 
 public class Database {
```
I would recommends to write simple unit tests for those methods, only to ensures the right behavior.

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/models/AppState.java:
```diff
@@ -23,4 +32,40 @@ public boolean getHasMoreGifs() {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.registerModule(new GuavaModule());
+        return mapper.readValue(json, AppState.class);
+    }
```
I would recommends to write simple unit tests for those methods, only to ensures the right behavior.

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/models/AppState.java:
```diff
@@ -23,4 +32,40 @@ public boolean getHasMoreGifs() {
+        }
+
+        return true;
+    }
```
The default equals method that is generated by Immutable library isn't enough? Why?

fpbitencourt replied:
O pmd obriga/recomenda sobrescrever esse método sempre que o equals é sobrecarregado, ainda que seja só pra chamar o super.

niltonvasques replied:
Hmm... Mas essa implementação é a mesma gerada pela lib immutables? Se sim, talvez seja adequado a gente dá um supress nesse warning do pmd. 
@raulccabreu pode dar um help com isso

raulccabreu replied:
If is the same why override? Finally is or isnt the same? The quality folder contains pmd supression file, but we have the annotation option too.

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/reducers/AppStateReducer.java:
```diff
@@ -23,6 +24,12 @@
+    public AppState stateLoaded(AppState oldState, AppState newState) {
+        return createImmutableAppBuilder(newState)
+                .build();
+    }
```
This new reduce action needs to be tested.

fpbitencourt committed:
Correções
1. pacotes duplicados
2. removendo asynctask para evitar vazamento de memória
3. testes unitários para a conversões JSON de Gif e AppState

fpbitencourt committed:
1. Removido as implementações dos métodos equals e hashcode em Gif e Appstate
2. Adicionados testes em DataManagerTest
3. Criado testes para as transformações Json/Objeto em Gif e AppState
4. Criado teste de redução da ação LOADED_APP_STATE

fpbitencourt committed:
removido --debug da configuração drone

fpbitencourt committed:
ajustes

raulccabreu commented:
@fpbitencourt rebase with master and tests will pass.

fpbitencourt committed:
Merge branch 'master' into integrating_with_no_sql

fpbitencourt committed:
integração com a master

raulccabreu commented on app/src/androidTest/java/br/com/catbag/gifreduxsample/asyncs/data/DataManagerTest.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private Map<String, Gif> buildGifs(int amount) {
+        Map<String, Gif> gifs = new HashMap<>();
```
Mesmo que seu teste não esteja quebrando com HashMap, outro teste futuro pode quebrar, então é bom manter LinkedHashMap.

raulccabreu commented on app/src/main/java/br/com/catbag/gifreduxsample/MyApp.java:
```diff
@@ -19,24 +20,37 @@
     }
 
+    private void initializeDataManager() {
+        mDataManager = new DataManager(getBaseContext());
```
Acho que mDataManager não precisa ser membro da classe, usa uma variável local lá no initializeFluxxan.

raulccabreu commented on app/src/main/java/br/com/catbag/gifreduxsample/middlewares/PersistenceMiddleware.java:
```diff
@@ -0,0 +1,53 @@
+ */
+
+public class PersistenceMiddleware extends BaseMiddleware<AppState> implements
+        StateListener<AppState> {
```
No futuro o middleware vai ter um método "before" e "after" e vai deixar de ser um stateListener. Eu abriria uma issue pra mudar isso na Fluxxan.

raulccabreu reviewed (changes requested):
Uma revisão adiantada...

fpbitencourt committed:
correções

fpbitencourt committed:
clean imports e refactor

fpbitencourt committed:
mockando database para o middleware de persistência

raulccabreu commented:
Approved 💯!

niltonvasques commented:
A lot of errors appear in the latest build on drone: 

```
br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendDownloadFailureAction STANDARD_ERROR
    Exception in thread "Thread-18" java.lang.UnsatisfiedLinkError: no snappydb-native in java.library.path
    	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
    	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
    	at java.lang.System.loadLibrary(System.java:1122)
    	at com.snappydb.internal.DBImpl.<clinit>(DBImpl.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendPauseAction STANDARD_ERROR
    Exception in thread "Thread-20" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendLoadListAction STANDARD_ERROR
    Exception in thread "Thread-22" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenCompleteAppFlowActions STANDARD_ERROR
    Exception in thread "Thread-24" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendLoadListWithANonEmptyListAction STANDARD_ERROR
    Exception in thread "Thread-26" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendManyActions STANDARD_ERROR
    Exception in thread "Thread-28" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendPlayAction STANDARD_ERROR
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)
    Exception in thread "Thread-30" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendDownloadSuccessAction STANDARD_ERROR
    Exception in thread "Thread-32" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.GifListReducerTest > whenSendDownloadStartedAction STANDARD_ERROR
    Exception in thread "Thread-34" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
    	at java.lang.Thread.run(Thread.java:745)

br.com.catbag.gifreduxsample.reducers.reducers.AppStateReducerTest > initialAppState STANDARD_ERROR
    Exception in thread "Thread-36" java.lang.NoClassDefFoundError: Could not initialize class com.snappydb.internal.DBImpl
    	at com.snappydb.DBFactory.open(DBFactory.java:40)
    	at com.snappydb.DBFactory.open(DBFactory.java:65)
    	at com.snappydb.DBFactory.open(DBFactory.java:77)
    	at br.com.catbag.gifreduxsample.asyncs.data.storage.Database.saveAppState(Database.java:35)
    	at br.com.catbag.gifreduxsample.asyncs.data.DataManager$SaveAppStateRunnable.run(DataManager.java:164)
```

raulccabreu commented:
And the test passed? omg

niltonvasques commented:
Weird behavior! However the check task, that also run unit tests, have passed without errors.

fpbitencourt committed:
desregistrando os middlewares nos testes de Reducers e desregistrando o PersistenceMiddleware para ouvir as mudanças do appstate

raulccabreu commented:
Tried a second build/test and the same errors occurs.

fpbitencourt commented:
É um problema da RobolectricTestRunner com códigos nativos. 

Desregistrei os middlewares e fiz PersistenceMiddleware deixar de ouvir as mudanças do appstate para os testes do reducer e tá tudo certo agora.

niltonvasques commented:
Coverage has increased from 69% to 74% :dancer: :100: :smiley_cat:

fpbitencourt committed:
desregistrando o PersistenceMiddleware para ouvir as mudanças do appstate no GifListActivityTest

fpbitencourt committed:
Merge branch 'master' into integrating_with_no_sql
